tax/scenario.go

```diff
@@ -0,0 +1,114 @@
+package tax
+
+import (
+	"github.com/invopop/gobl/cbc"
+	"github.com/invopop/validation"
+)
+
+// ScenarioSet is a collection of tax scenarios for a given schema that can be used to
+// determine special codes or notes that need to be included in the final document.
+type ScenarioSet struct {
+	// Partial or complete schema URL for the document type
+	Schema string `json:"schema" jsonschema:"title=Schema"`
+	// List of scenarios for the schema
+	List []*Scenario `json:"list" jsonschema:"title=List"`
+}
+
+// Scenario is used to describe a tax scenario of a document based on the combination
+// of document type and tag used.
+type Scenario struct {
+	// Type of document, if present.
+	Types []cbc.Key `json:"type,omitempty" jsonschema:"title=Type"`
+
+	// Tag that was applied to the document
+	Tags []cbc.Key `json:"tags,omitempty" jsonschema:"title=Tag"`
+
+	// A note to be added to the document if the scenario is applied.
+	Note *cbc.Note `json:"note,omitempty" jsonschema:"title=Note"`
+
+	// Any additional local meta data that may be useful in integrations.
+	Meta cbc.Meta `json:"meta,omitempty" jsonschema:"title=Meta"`
+}
+
+// ScenarioSummary is the result after running through a set of
+// scenarios and determining which combinations of Notes and Meta
+// are viable.
+type ScenarioSummary struct {
+	Notes []*cbc.Note
+	Meta  cbc.Meta
+}
+
+// Validate checks the scenario set for errors.
+func (ss *ScenarioSet) Validate() error {
+	err := validation.ValidateStruct(ss,
+		validation.Field(&ss.Schema, validation.Required),
+		validation.Field(&ss.List, validation.Required),
+	)
+	return err
+}
+
+// SummaryFor returns a summary by applying the scenarios to the
+// supplied document.
+func (ss *ScenarioSet) SummaryFor(docType cbc.Key, docTags []cbc.Key) *ScenarioSummary {
+	summary := &ScenarioSummary{
+		Notes: make([]*cbc.Note, 0),
+		Meta:  make(cbc.Meta),
+	}
+	for _, s := range ss.List {
+		if s.match(docType, docTags) {
+			if s.Note != nil {
+				summary.Notes = append(summary.Notes, s.Note)
+			}
+			for k, v := range s.Meta {
+				summary.Meta[k] = v
+			}
+		}
+	}
+	return summary
+}
+
+func (s *Scenario) match(docType cbc.Key, docTags []cbc.Key) bool {
+	if s.Types != nil {
```

Comment:
Would `Types` ever be `nil` given it's not a pointer type?

Reply:
(Would it not be an empty array if not specified?)

Reply:
In go, nil is the same as an empty array, but I agree, `len(s.Types) > 0` would be clearer!